rust/nasl-interpreter/src/built_in_functions/packet_forgery.rs

```diff
@@ -0,0 +1,2872 @@
+    };
+    pkt.set_next_level_protocol(IpNextHeaderProtocol::new(ip_p));
+
+    match register.named("ip_src") {
+        Some(ContextType::Value(NaslValue::String(x))) => {
+            match x.parse::<Ipv4Addr>() {
+                Ok(ip) => {
+                    pkt.set_source(ip);
+                }
+                Err(e) => {
+                    return Err(FunctionErrorKind::Diagnostic(
+                        format!("forge_ip_packet invalid ip_src: {}", e),
+                        Some(NaslValue::Null),
+                    ));
+                }
+            };
+            x.to_string()
+        }
+        _ => String::new(),
+    };
+
+    match register.named("ip_dst") {
+        Some(ContextType::Value(NaslValue::String(x))) => {
+            match x.parse::<Ipv4Addr>() {
+                Ok(ip) => {
+                    pkt.set_destination(ip);
+                }
+                Err(e) => {
+                    return Err(FunctionErrorKind::Diagnostic(
+                        format!("forge_ip_packet invalid ip_dst: {}", e),
+                        Some(NaslValue::Null),
+                    ));
+                }
+            };
+            x.to_string()
+        }
+        _ => {
+            match dst_addr.to_string().parse::<Ipv4Addr>() {
+                Ok(ip) => {
+                    pkt.set_destination(ip);
+                }
+                Err(e) => {
+                    return Err(FunctionErrorKind::Diagnostic(
+                        format!("forge_ip_packet invalid ip: {}", e),
+                        Some(NaslValue::Null),
+                    ));
+                }
+            };
+            dst_addr.to_string()
+        }
+    };
+
+    let ip_sum = match register.named("ip_sum") {
+        Some(ContextType::Value(NaslValue::Number(x))) => (*x as u16).to_be(),
+        _ => checksum(&pkt.to_immutable()),
+    };
+    pkt.set_checksum(ip_sum);
+
+    Ok(NaslValue::Data(buf))
+}
+
+/// Set element from a IP datagram. Its arguments are:
+///  
+/// - ip: IP datagram to set fields on
+/// - ip_hl: IP header length in 32 bits words, 5 by default
+/// - ip_id: datagram ID, random by default
+/// - ip_len: length of the datagram, 20 plus the length of the data
+/// - ip_off: fragment offset in 64 bits words, 0 by default
+/// - ip_p: IP protocol, 0 by default
+/// - ip_src: source address in ASCII, NASL will convert it into an integer in network order
+/// - ip_sum: packet header checksum, it will be computed by default
+/// - ip_tos: type of service field, 0 by default
+/// - ip_ttl: time to live field, 64 by default
+/// - ip_v: IP version, 4 by default
+///  
+/// Returns the modified IP datagram
+fn set_ip_elements<K>(
+    register: &Register,
+    _configs: &Context<K>,
+) -> Result<NaslValue, FunctionErrorKind> {
+    let mut buf = match register.named("ip") {
+        Some(ContextType::Value(NaslValue::Data(d))) => d.clone(),
+        _ => {
+            return Err(FunctionErrorKind::Diagnostic(
```

Comment:
is that really a diagnostic error? to me it seems more like a usage error.

Reply:
There is a `MissingArguments` error for such cases